specs/merge/p2p-interface.md

```diff
@@ -0,0 +1,111 @@
+# Ethereum Merge networking specification
+
+This document contains the networking specification for Ethereum 2.0 clients added during the Merge deployment.
+
+The specification of these changes continues in the same format as the network specifications of previous upgrades, and assumes them as pre-requisite. This document should be viewed as additive to the documents from [Phase 0](../phase0/p2p-interface.md) and from [Altair](../altair/p2p-interface.md)
+and will be referred to as the "Phase 0 document" and "Altair document" respectively, hereafter.
+Readers should understand the Phase 0 and Altair documents and use them as a basis to understand the changes outlined in this document.
+
+## Table of contents
+
+<!-- TOC -->
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+  - [Warning](#warning)
+- [Modifications in the Merge](#modifications-in-the-merge)
+  - [The gossip domain: gossipsub](#the-gossip-domain-gossipsub)
+    - [Topics and messages](#topics-and-messages)
+      - [Global topics](#global-topics)
+        - [`beacon_block`](#beacon_block)
+    - [Transitioning the gossip](#transitioning-the-gossip)
+  - [The Req/Resp domain](#the-reqresp-domain)
+    - [Messages](#messages)
+      - [BeaconBlocksByRange v2](#beaconblocksbyrange-v2)
+      - [BeaconBlocksByRoot v2](#beaconblocksbyroot-v2)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Warning
+
+This document is currently illustrative for early Merge testnets and some parts are subject to change.
+Refer to the note in the [validator guide](./validator.md) for further details.
+
+# Modifications in the Merge
+
+## The gossip domain: gossipsub
+
+Some gossip meshes are upgraded in the Merge to support upgraded types.
+
+### Topics and messages
+
+Topics follow the same specification as in prior upgrades.
+All topics remain stable except the beacon block topic which is updated with the modified type.
+
+The specification around the creation, validation, and dissemination of messages has not changed from the Phase 0 and Altair documents.
+
+The derivation of the `message-id` remains stable.
+
+The new topics along with the type of the `data` field of a gossipsub message are given in this table:
+
+| Name | Message Type |
+| - | - |
+| `beacon_block` | `SignedBeaconBlock` (modified) |
+
+Note that the `ForkDigestValue` path segment of the topic separates the old and the new `beacon_block` topics.
+
+#### Global topics
+
+The Merge changes the type of the global beacon block topic.
+
+##### `beacon_block`
+
+The existing specification for this topic does not change from prior upgrades,
```

Comment:
I think we can introduce additional gossip topic validation steps, to filter bad execution-payloads, without requiring any execution-layer storage:
- [reject] Check `gas_used <= gas_limit`
- [reject] Check `gas_limit <= GAS_SANITY_LIMIT`
- [reject] `timestamp != genesis_time + slot * SECONDS_PER_SLOT`
- [reject] `block_hash != parent_hash`
- [reject] Check transactions data size

And if we want to track the execution chain history (or maybe just a cache of the latest hashes and numbers?)
- [ignore] Check `parent_hash` is known
- [reject] Check `block_number == previous_block_number + 1`

Reply:
yeah, good call. also checking what other low-cost validations geth does today

Reply:
The current devp2p gossip validations can be found here -- https://github.com/ethereum/devp2p/blob/master/caps/eth.md#block-encoding-and-validity

All cheap checks should be replicated here as much as is possible. Ideally, we do not have to add a `validateHeader` call to the execution-engine. I would argue it's even okay for a "cheap" condition or two to be dropped if it is too difficult for the consensus layer to validate. The proposer signature (like the proof of work) is the most important initial validation. All others are nice-to-have

Reply:
> Check parent_hash is known

comes trivially from knowing and fully processing the `beacon_block` parent. Will keep that embedded in that validation.

> [reject] Check block_number == previous_block_number + 1

going to avoid adding state/cache here for now